### PR TITLE
feat: initialize enterprise metadata in HandlerContext; complete auto-apply business logic

### PR DESCRIPTION
**Pull request by adamstankiewicz**
**Jira:**
[ENT-9629](https://2u-internal.atlassian.net/browse/ENT-9629)

**Description:**

* Converted additional `HandlerContext` class attributes as private, with `@property` getter functions.
* Refactors `initialize_common_context_data` as `_initialize_common_context_data`:
  * Extended to optionally support both `enterprise_customer_uuid` and `enterprise_customer_slug`.
  * Extended to include an attribute for `enterprise_features` (returned by `enterprise-learner` API, described below).
  * Initializes context `data` with enterprise customer metadata (mostly parity with transforms on the existing Learner Portal request) via the `enterprise-learner` LMS API.
    * If no slug/uuid is provided, resolves to the request user's `active=True` enterprise customer user.
    * If only slug is provided, finds linked enterprise customer matching the requested slug and sets the uuid.
    * If only uuid is provided, finds linked enterprise customer matching the requested uuid and sets the slug.
    * If uuid and/or slug is provided, but no matching linked enterprise customer is found, falls back to request the enterprise customer metadata as a staff user (if applicable).
* Completes the auto-apply business logic (to be tested at the viewset layer in separate ticket) by extending the conditional check to include whether the enterprise customer has an `identity_provider` configured.
* Integrates with `self.lms_user_api_client.get_default_enterprise_enrollment_intentions_learner_status` in `BaseLearnerPortalHandler`.
* Integrates with `self.lms_user_api_client.get_enterprise_course_enrollments` in `DashboardHandler`.
* Tests.

**Merge checklist:**
- [ ] `./manage.py makemigrations` has been run
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.

**Post merge:**
- [ ] Ensure that your changes went out to the stage instance
- [ ] Deploy to prod instance

### CODE REVIEW

File: enterprise_access/apps/bffs/context.py [hunk 1]
```diff
@@ -43,23 +54,203 @@ def request(self):
...
+        )
+        enterprise_customer = self._determine_enterprise_customer_for_display(
+            active_enterprise_customer=active_enterprise_customer,
+            requested_enterprise_customer=requested_enterprise_customer,
+            staff_enterprise_customer=staff_enterprise_customer,
+        )
+
+        # Update the context data with the enterprise customer user information
+        self.data = {
+            **self.data,
```
Comment (adamstankiewicz) [inline enterprise_access/apps/bffs/context.py:192]: TODO: use `self.data.update({ ... })` instead

File: enterprise_access/apps/bffs/handlers.py [hunk 2]
```diff
@@ -69,29 +73,94 @@ def load_and_process(self):
...
+
             # Retrieve and process subscription licenses. Handles activation and auto-apply logic.
-            # TODO: retrieve enterprise customer metadata,ENT-9629
-            # self.load_enterprise_customer()
             self.load_and_process_subscription_licenses()
 
-            # Retrieve default enterprise courses and enroll in the redeemable ones
-            self.load_default_enterprise_courses()
-            self.enroll_in_redeemable_default_courses()
+            # # Retrieve default enterprise courses and enroll in the redeemable ones
```
Comment (adamstankiewicz) [inline enterprise_access/apps/bffs/handlers.py:82]: TODO: double comment

File: enterprise_access/apps/bffs/handlers.py [hunk 3]
```diff
@@ -69,29 +73,94 @@ def load_and_process(self):
...
+            self.load_default_enterprise_enrollment_intentions()
+            self.enroll_in_redeemable_default_enterprise_enrollment_intentions()
         except Exception as e:  # pylint: disable=broad-exception-caught
             logger.exception("Error loading learner portal handler")
             self.add_error(
                 user_message="An error occurred while loading and processing common learner logic.",
                 developer_message=f"Error: {e}",
             )
 
+    def transform_enterprise_customers(self):
```
Comment (adamstankiewicz) [inline enterprise_access/apps/bffs/handlers.py:92]: TODO: i believe this is missing transforming the `staff_enterprise_customer` property

File: enterprise_access/apps/bffs/context.py [hunk 4]
```diff
@@ -43,23 +54,203 @@ def request(self):
...
+                enterprise_customer_user
+                for enterprise_customer_user in enterprise_customer_users
+                if enterprise_customer_user.get('active', False)
+            ),
+            None
+        )
+        active_enterprise_customer = (
+            active_enterprise_customer_user.get('enterprise_customer')
+            if active_enterprise_customer_user else None
+        )
```
Comment (brobro10000) [inline enterprise_access/apps/bffs/context.py:148]: [nit/non-blocking] This functions doing a lot. I wonder if its worth chunking up some of the logic into a separate function?
Ex. `get_enterprise_customer_users_data`, `get_active_enterprise_customer`

File: enterprise_access/apps/bffs/context.py [hunk 5]
```diff
@@ -43,23 +54,203 @@ def request(self):
...
+            try:
+                staff_enterprise_customer = self.lms_api_client.get_enterprise_customer_data(
+                    enterprise_customer_uuid=self.enterprise_customer_uuid,
+                    enterprise_customer_slug=self.enterprise_customer_slug,
+                )
+            except Exception as e:  # pylint: disable=broad-except
+                self.add_error(
+                    user_message='Error retrieving enterprise customer data',
+                    developer_message=str(e)
+                )
```
Comment (brobro10000) [inline enterprise_access/apps/bffs/context.py:177]: [nit/non-blocking] Another opportunity to chunk logic into a separate function here.
Ex. `get_staff_enterprise_customer`

File: enterprise_access/apps/bffs/context.py [hunk 6]
```diff
@@ -43,23 +54,203 @@ def request(self):
...
+
+        slug_matches_active_enterprise_customer = (
+            active_enterprise_customer and active_enterprise_customer.get('slug') == self.enterprise_customer_slug
+        )
+        uuid_matches_active_enterprise_customer = (
+            active_enterprise_customer and active_enterprise_customer.get('uuid') == self.enterprise_customer_uuid
+        )
+        request_matches_active_enterprise_customer = (
+            slug_matches_active_enterprise_customer or uuid_matches_active_enterprise_customer
+        )
```
Comment (brobro10000) [inline enterprise_access/apps/bffs/context.py:223]: [nit/non-blocking] could this be extracted into a helper function
Ex. `validate_active_enterprise_customer`

**Review by brobro10000 — APPROVED** [review]
Great job. A couple of nits that can be deferred or ignored. Approving to unblock.

**Review by brobro10000 — APPROVED** [review]

File: enterprise_access/apps/bffs/handlers.py [hunk 7]
```diff
@@ -69,29 +73,97 @@ def load_and_process(self):
...
+        """
+        Transform enterprise customer metadata retrieved by self.context.
+        """
+        if enterprise_customer := self.context.data.get('enterprise_customer'):
```
Comment (iloveagent57) [inline enterprise_access/apps/bffs/handlers.py:96]: nit: `if self.context.enterprise_customer`

Comment (iloveagent57) [reply enterprise_access/apps/bffs/handlers.py:96]: actually, how about a loop like this?
```python
for customer_record_key in ('enterprise_customer', 'active_enterprise_customer', 'staff_enterprise_customer'):
    if not customer_record := self.context.data.get(customer_record_key):
        continue
    self.context.data[customer_record_key] = self.transform_enterprise_customer(customer_record)
```

File: enterprise_access/apps/bffs/handlers.py [hunk 8]
```diff
@@ -69,29 +73,97 @@ def load_and_process(self):
...
+    def transform_enterprise_customer_user(self, enterprise_customer_user):
+        """
+        Transform the enterprise customer user data.
+
+        Args:
+            enterprise_customer_user: The enterprise customer user data.
+        Returns:
+            The transformed enterprise customer user data.
+        """
+        enterprise_customer = enterprise_customer_user.get('enterprise_customer')
```
Comment (iloveagent57) [inline enterprise_access/apps/bffs/handlers.py:119]: The `enterprise_customer` dict in the ECU response payload won't be "full", right? Does that matter here?

File: enterprise_access/apps/api_client/lms_client.py [hunk 9]
```diff
@@ -406,13 +415,67 @@ class LmsUserApiClient(BaseUserApiClient):
...
+                    initial_response_data = data
+
+                # Collect results from the current page
+                results.extend(data.get('results', []))
+
+                # If pagination is enabled, continue with the next page; otherwise, break
+                next_url = data.get('next') if traverse_pagination else None
+
+            consolidated_response = {
+                **initial_response_data,
```
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply enterprise_access/apps/api_client/lms_client.py:464]: Correct, and then the subsequent lines below overrides relevant fields like `next`, `prev`, etc. to reflect the traversed pagination results being a single page.

The primary reason to spread `**initial_response_data` is bring along any other existing pagination fields unrelated to page length, i.e. I believe `**initial_response_data` brings along `start: 0` as well without needing to explicitly define it. The `start` value from the initial response is unchanged regardless of `traverse_pagination`.

File: enterprise_access/apps/bffs/context.py [hunk 10]
```diff
@@ -43,23 +54,224 @@ def request(self):
...
             enterprise_uuid_post_param = self.request.data.get('enterprise_customer_uuid')
+            enterprise_slug_post_param = self.request.data.get('enterprise_customer_slug')
 
         enterprise_customer_uuid = enterprise_uuid_query_param or enterprise_uuid_post_param
-        if enterprise_customer_uuid:
-            self.enterprise_customer_uuid = enterprise_customer_uuid
-        else:
-            raise ValueError("enterprise_customer_uuid is required for this request.")
+        self._enterprise_customer_uuid = enterprise_customer_uuid
+        enterprise_customer_slug = enterprise_slug_query_param or enterprise_slug_post_param
```
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply enterprise_access/apps/bffs/context.py:100]: Can you expand on what the proposed `assert` would be enforcing? Would this suggest that the `enterprise_customer_slug` is required?

This PR refactors to make both the `enterprise_customer_uuid` and `enterprise_customer_slug` params optional. When neither is provided, it returns the request user's linked enterprise customer with `active=True` (similar to some existing frontend business logic).

[the remark this answers is not in the document]
Comment (iloveagent57) [reply enterprise_access/apps/bffs/context.py:100]: Ah, ok, that makese sense. Nix this one.

File: enterprise_access/apps/bffs/context.py [hunk 11]
```diff
@@ -43,23 +54,224 @@ def request(self):
...
+                for enterprise_customer_user in enterprise_customer_users
+                if self._enterprise_customer_matches_slug_or_uuid(enterprise_customer_user.get('enterprise_customer'))
+            ),
+            None
+        )
+
+        # If no enterprise customer user is found for the requested customer (i.e., request user not explicitly
+        # linked), but the request user is staff, attempt to retrieve enterprise customer metadata from the
+        # `/enterprise-customer/` LMS API endpoint instead.
+        staff_enterprise_customer = self._get_staff_enterprise_customer(enterprise_customer_user_for_requested_customer)
```
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply enterprise_access/apps/bffs/context.py:150]: Agreed, good suggestion for clean up to make it easier to grok.

File: enterprise_access/apps/bffs/handlers.py (hunk 7 above)
Comment (adamstankiewicz) [reply enterprise_access/apps/bffs/handlers.py:96]: I like the loop suggestion to make the implementation more DRY. I believe the conditional logic around `if enterprise_customer_users := self.context.data.get('all_linked_enterprise_customer_users'):` will still need to remain, though.

```python
def transform_enterprise_customers(self):
    """
    Transform enterprise customer metadata retrieved by self.context.
    """
    for customer_record_key in ('enterprise_customer', 'active_enterprise_customer', 'staff_enterprise_customer'):
        if not (customer_record := self.context.data.get(customer_record_key)):
            continue
        self.context.data[customer_record_key] = self.transform_enterprise_customer(customer_record)

    if enterprise_customer_users := self.context.data.get('all_linked_enterprise_customer_users'):
        self.context.data['all_linked_enterprise_customer_users'] = [
            self.transform_enterprise_customer_user(enterprise_customer_user)
            for enterprise_customer_user in enterprise_customer_users
        ]
```

File: enterprise_access/apps/bffs/handlers.py (hunk 8 above)
Comment (adamstankiewicz) [reply enterprise_access/apps/bffs/handlers.py:119]: The `enterprise-learner` API uses the same `EnterpriseCustomerSerializer` that the list/detail enterprise customer API uses; it's the "full" metadata. This is the same enterprise customer metadata currently consumed by the frontends (both learner/admin).

That said, when cross-checking to verify, I noticed a discrepancy between `admin_users` list in the `enterprise-learner` API (this data source) vs. the `enterprise-customer` detail API. The `enterprise-learner` API _always_ denotes `admin_users` as an empty list, even if the customer has admin users...

I filed a [bug ticket](https://2u-internal.atlassian.net/browse/ENT-9741) and fixed the bug through this PR: https://github.com/openedx/edx-enterprise/pull/2287

File: enterprise_access/apps/api_client/lms_client.py [hunk 12]
```diff
@@ -64,21 +63,31 @@ def enterprise_group_members_endpoint(self, group_uuid):
...
         try:
-            endpoint = f'{self.enterprise_customer_endpoint}{enterprise_customer_uuid}/'
             response = self.client.get(endpoint, timeout=settings.LMS_CLIENT_TIMEOUT)
             response.raise_for_status()
-            return response.json()
+            if enterprise_customer_uuid:
+                # If we're fetching by UUID, we expect a single result
+                return response.json()
+            # If we're fetching by slug, we expect a list of results
+            return response.json().get('results', [])[0]
```
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply enterprise_access/apps/api_client/lms_client.py]: Pushed up a commit with a refactor. Ended up with:

```python
payload = response.json()
if results := payload.get('results'):
  return results[0]
return payload
```

**Review by iloveagent57 — APPROVED** [review]

**Pull request merged**